workers/unity/Packages/io.improbable.gdk.debug/WorkerInspector/Codegen/PaginatedListView.cs

```diff
@@ -0,0 +1,151 @@
+            template.CloneTree(this);
+
+            this.Q<Label>(name: "list-name").text = label;
+            container = this.Q<VisualElement>(className: "user-defined-type-container-data");
+
+            controlsContainer = this.Q<VisualElement>(className: "paginated-list-controls");
+            pageCounter = this.Q<Label>(name: "page-counter");
+
+            backButton = this.Q<Button>(name: "back-button");
+            backButton.text = "<";
+            backButton.clickable.clicked += () => ChangePageCount(-1);
+
+            forwardButton = this.Q<Button>(name: "forward-button");
+            forwardButton.text = ">";
+            forwardButton.clickable.clicked += () => ChangePageCount(1);
+
+            elementPool = new Pool(makeElement);
+        }
+
+        public void Update(List<TData> newData)
+        {
+            data = newData;
+
+            if (data.Count == 0)
+            {
+                controlsContainer.AddToClassList("hidden");
+            }
+            else
+            {
+                controlsContainer.RemoveFromClassList("hidden");
+            }
+
+            CalculatePages();
+            RefreshView();
+        }
+
+        internal void ChangePageCount(int diff)
+        {
+            currentPage += diff;
+            currentPage = Mathf.Clamp(currentPage, 0, numPages - 1);
+            CalculatePages();
+            RefreshView();
+        }
+
+        private void CalculatePages()
+        {
+            numPages = (int) Math.Ceiling((double) data.Count / elementsPerPage);
+            numPages = Mathf.Clamp(numPages, 1, numPages);
+            currentPage = Mathf.Clamp(currentPage, 0, numPages - 1);
+
+            pageCounter.text = $"{currentPage + 1}/{numPages}";
+        }
+
+        private void RefreshView()
+        {
+            // Calculate slice of list to be rendered.
+            var firstIndex = currentPage * elementsPerPage;
+            var length = Math.Min(elementsPerPage, data.Count - firstIndex);
+
+            // If the child count is the same, don't adjust it.
+            // If the child count is less, add the requisite number.
+            // If the child count is more, pop elements off the end.
+
+            var diff = container.childCount - length;
+
+            if (diff > 0)
+            {
+                for (var i = 0; i < diff; i++)
+                {
+                    var element = container.ElementAt(container.childCount - 1);
+                    container.RemoveAt(container.childCount - 1);
+                    elementPool.Return((TElement) element);
+                }
+            }
+            else if (diff < 0)
+            {
+                for (var i = diff; i < 0; i++)
+                {
+                    container.Add(elementPool.GetOrCreate());
+                }
+            }
+
+            // At this point, container.Children() has the same length as the slice.
+            var j = 0;
```

Comment:
`var elementIndex= firstIndex;` and then increase that index, instead of using `j`

Reply:
Ah yeah this got missed in a refactor, used to be a `for` rather than `foreach`